src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -96,7 +106,62 @@ void setVarargs(JCExpression tree, FunctionType type) {
         }
     }
 
-    private JCTree opToTree(Op op) {
+    public JCTree.JCMethodDecl transformFuncOpToAST(CoreOp.FuncOp funcOp, Name methodName) {
+        Assert.check(funcOp.body().blocks().size() == 1);
+
+        var paramTypes = List.of(crSym.opFactoryType, crSym.typeElementFactoryType);
+        var mt = new Type.MethodType(paramTypes, crSym.opType, List.nil(), syms.methodClass);
+        ms = new Symbol.MethodSymbol(PUBLIC | STATIC | SYNTHETIC, methodName, mt, currClassSym);
+        currClassSym.members().enter(ms);
+
+        for (int i = 0; i < funcOp.parameters().size(); i++) {
+            valueToTree.put(funcOp.parameters().get(i), treeMaker.Ident(ms.params().get(i)));
+        }
+
+        final MethodRef BLOCK_BUILDER_OP = MethodRef.method(Block.Builder.class, "op",
+                Op.Result.class, Op.class);
+        final MethodRef BLOCK_BUILDER_PARAM = MethodRef.method(Block.Builder.class, "parameter",
+                Block.Parameter.class, TypeElement.class);
+        Set<Value> vals = funcOp.traverse(new HashSet<>(), (l, ce) -> {
+           if (ce instanceof CoreOp.InvokeOp invokeOp && (invokeOp.invokeDescriptor().equals(BLOCK_BUILDER_OP)
+                   || invokeOp.invokeDescriptor().equals(BLOCK_BUILDER_PARAM))) {
+               l.add(invokeOp.result());
+           }
+            return l;
+        });
+        Map<Value, Node<Value>> prunedGraphs = prunedExpressionGraphs(funcOp, vals);
+        java.util.List<Node<Value>> prunedRootGraphs = prunedGraphs.values().stream()
+                .filter(n -> n.value() instanceof Op.Result opr &&
+                        switch (opr.op()) {
+                            // Variable declarations modeling local variables
+                            case CoreOp.VarOp vop -> vop.operands().get(0) instanceof Op.Result;
+                            // An operation result with no uses or more than one
+                            default -> opr.uses().size() != 1 || vals.contains(opr);
+                        })
+                .toList();
+
+        var stats = new ListBuffer<JCTree.JCStatement>();
+        // instead, we will traverse the root expr graphs
```

Comment:
instead of what? Perhaps say "Traverse the root expr graphs instead of traversing the model"